File: Formant Analyzer/FormantPlotter/SpeechAnalyzer.m
//
//  SpeechAnalyzer.m
//  FormantPlotter
//
//  Created by William Entriken on 1/19/15.
//  Copyright (c) 2015 William Entriken. All rights reserved.
//

#import "SpeechAnalyzer.h"
#import <complex.h>

// A few constants to be used in LPC and Laguerre algorithms.
#define ORDER 20
#define EPS 2.0e-6
#define EPSS 1.0e-7
#define MR 8
#define MT 10
#define MAXIT (MT*MR)

@interface SpeechAnalyzer()
@property (nonatomic) NSData *int16Samples;
@property (nonatomic) NSData *int16SamplesDecimated;
@property (nonatomic) NSRange strongSignalRangeCached;
@property (nonatomic) NSArray *cleanFormantsCached;
- (NSRange)strongSignalRange;
@end

@implementation SpeechAnalyzer

+ (SpeechAnalyzer *)analyzerWithData:(NSData *)int16Samples
{
    SpeechAnalyzer *retval = [[SpeechAnalyzer alloc] init];
    [retval loadData:int16Samples];
    return retval;
}

- (id)init
{
    self = [super init];
    if (self) {
        self.strongSignalRangeCached = NSMakeRange(0, 0);
    }
    return self;
}

- (NSNumber *)totalSamples
{
    return @(self.int16Samples.length / 2);
}

// Trim quiet parts at the ends of our signal.
// Our signal is divided into 300 chunks, and energy is computed for each.
// Leading and trailing chunks are excluded with 10dB less than the maximum energy.
// The range of the remaining signal is returned.
- (NSRange)strongSignalRange
{
    if (self.strongSignalRangeCached.length) {
        return self.strongSignalRangeCached;
    }
    
    long numChunks = 300;
    long chunkEnergy, chunkEnergyThreshold;
    long maxChunkEnergy = 0;
    long startSample = 0, endSample = INT_MAX;
    long chunkSize;
    short int *dataBuffer = (short int*)self.int16Samples.bytes;
    chunkSize = self.int16Samples.length / sizeof(short int) / numChunks;

    // Find the chunk with the most energy and set energy threshold
    for (int chunkIdx = 0; chunkIdx < numChunks; chunkIdx++) {
        chunkEnergy = 0;
        for (int chunkSampleIdx = 0; chunkSampleIdx < chunkSize; chunkSampleIdx++) {
            chunkEnergy += dataBuffer[chunkIdx * chunkSize + chunkSampleIdx] * dataBuffer[chunkIdx * chunkSize + chunkSampleIdx] / 1000;
        }
        maxChunkEnergy = MAX(maxChunkEnergy, chunkEnergy);
    }
    
    chunkEnergyThreshold = maxChunkEnergy / 10;
    
    // Find starting sample meeting minimum energy threshold
    startSample = 0;
    for (int chunkIdx = 0; chunkIdx < numChunks; chunkIdx++) {
        chunkEnergy = 0;
        for (int chunkSampleIdx = 0; chunkSampleIdx < chunkSize; chunkSampleIdx++) {
            chunkEnergy += dataBuffer[chunkIdx * chunkSize + chunkSampleIdx] * dataBuffer[chunkIdx * chunkSize + chunkSampleIdx] / 1000;
        }
        if (chunkEnergy > chunkEnergyThreshold) {
            startSample = chunkIdx * chunkSize;
            break;
        }
    }
    
    // Find ending sample meeting minimum energy threshold
    endSample = self.int16Samples.length / sizeof(short int);
    for (long chunkIdx = numChunks-1; chunkIdx >= 0; chunkIdx--) {
        chunkEnergy = 0;
        for (int chunkSampleIdx = 0; chunkSampleIdx < chunkSize; chunkSampleIdx++) {
            chunkEnergy += dataBuffer[chunkIdx * chunkSize + chunkSampleIdx] * dataBuffer[chunkIdx * chunkSize + chunkSampleIdx] / 1000;
        }
        if (chunkEnergy > chunkEnergyThreshold) {
            endSample = (chunkIdx + 1) * chunkSize - 1;
            break;
        }
    }
    
    self.strongSignalRangeCached = NSMakeRange(startSample, endSample-startSample);
    return self.strongSignalRangeCached;
}

// Removes 15% off each end of a range
- (NSRange)truncateRangeTails:(NSRange)range
{
    NSUInteger newLength = range.length * 0.7;
    NSUInteger newLocation = range.location + range.length * 0.15;
    return NSMakeRange(newLocation, newLength);
}

/*
 * This takes the isolated vowel range of the sound signal and reduces the sampling rate.
 * A simple, one-pass algorithm is used.
 * Because sampling rate is 44.1kHz and human speech formants are < 5kHz, quality not needed here.
 */
- (NSData *)int16SamplesDecimated
{
    if (!_int16SamplesDecimated) {
        int decimationFactor = 4; // input variable, 11025 Hz sample rate
        short int *inputDataBuffer = (short int*)self.int16Samples.bytes;
        NSRange vowelRange = [self vowelRange];
        long vowelStartIdx = vowelRange.location;
        long vowelEndIdx = vowelRange.location + vowelRange.length;
        long outputSamples = (vowelEndIdx - vowelStartIdx) / decimationFactor;
        
        NSMutableData *workingData = [NSMutableData data];
        long outputIndex;
        for (outputIndex = 0; outputIndex < outputSamples; outputIndex++) {
            [workingData appendBytes:&inputDataBuffer[4*outputIndex + vowelStartIdx] length:sizeof(short int)];
        }
        _int16SamplesDecimated = workingData;
    }
    return _int16SamplesDecimated;
}

- (void)loadData:(NSData *)int16Samples
{
    self.int16Samples = int16Samples;
    self.strongSignalRangeCached = NSMakeRange(0, 0);
}

/**
 * Reduce horizontal resolution of signal for plotting, returns NO on error
 */
- (NSArray *)downsampleToSamples:(int)samples
{
    short int *dataBuffer = (short int*)self.int16Samples.bytes;
//    NSRange strongRange = [self strongSignalRange];
    NSRange strongRange = NSMakeRange(0, self.totalSamples.doubleValue);
    long strongStartIdx = strongRange.location;
    long strongEndIdx = strongRange.location + strongRange.length;
    long chunkSamples = (strongEndIdx - strongStartIdx) / 400;
    NSMutableArray *plottableValuesHigh = [NSMutableArray array];
    
    for (long chunkIdx=0; chunkIdx<400; chunkIdx++) {
        long chunkMaxValue = 0;
        for (long j=0; j<chunkSamples; j++) {
            long dataBufferIdx = j + strongStartIdx + chunkIdx*chunkSamples;
            chunkMaxValue = MAX(chunkMaxValue, dataBuffer[dataBufferIdx]);
        }
        [plottableValuesHigh addObject:@(chunkMaxValue)];
    }
    return plottableValuesHigh;
}

/**
 * Find start and finish points representing vowel signal in speech data
 */
- (NSRange)vowelRange
{
    NSRange range = [self strongSignalRange];
    range = [self truncateRangeTails:range];
    return range;
}

/**
 * Find linear prediction coefficients using an iterative procedure of Levinson
 */
- (NSArray *)lpcCoefficients
{
    NSData *decimatedBufferData = self.int16SamplesDecimated;
    short int *decimatedBuffer = (short int *)decimatedBufferData.bytes;
    long decimatedBufferSamples = decimatedBufferData.length / sizeof(short int);
    
    // Find ORDER+1 autocorrelation coefficient
    NSMutableArray *Rxx = [NSMutableArray arrayWithCapacity:ORDER + 1];
    NSMutableArray *pCoeff = [NSMutableArray arrayWithCapacity:ORDER + 1];
    
    // Find all the correlation coefficients.
    for (int delayIdx = 0; delayIdx <= ORDER; delayIdx++) {
        double corrSum = 0;
        for (int dataIdx = 0; dataIdx < (decimatedBufferSamples - delayIdx); dataIdx++) {
            corrSum += (double)decimatedBuffer[dataIdx] * decimatedBuffer[dataIdx + delayIdx];
        }
        Rxx[delayIdx] = @(corrSum);
    }
    
    // Now solve for the predictor coefficients
    double pError = ((NSNumber *)Rxx[0]).doubleValue;      // initialize error to total power
    pCoeff[0] = @(1.0);                                    // first coefficient must be = 1 (perfect autocorrelation)
    
    // For each coefficient in turn
    for (int k = 1 ; k <= ORDER ; k++) {
        // find next reflection coeff from pCoeff[] and Rxx[]
        double rcNum = 0;
        for (int i = 1 ; i <= k ; i++) {
            rcNum -= ((NSNumber *)pCoeff[k-i]).doubleValue * ((NSNumber *)Rxx[i]).doubleValue;
        }
        
        pCoeff[k] = @(rcNum/pError);
        
        // perform recursion on pCoeff[]
        for (int i = 1 ; i <= k/2 ; i++) {
            double pci  = ((NSNumber *)pCoeff[i]).doubleValue + ((NSNumber *)pCoeff[k]).doubleValue * ((NSNumber *)pCoeff[k-i]).doubleValue;
            double pcki = ((NSNumber *)pCoeff[k-i]).doubleValue + ((NSNumber *)pCoeff[k]).doubleValue * ((NSNumber *)pCoeff[i]).doubleValue;
            pCoeff[i] = @(pci);
            pCoeff[k-i] = @(pcki);
        }
        
        // calculate residual error
        pError = pError * (1.0 - ((NSNumber *)pCoeff[k]).doubleValue * ((NSNumber *)pCoeff[k]).doubleValue);
    }
    
    // WE hack: fix for bad coefficients, undocumented
    for (int i = 0; i <= ORDER; i++) {
        if (isnan(((NSNumber *)pCoeff[i]).doubleValue)) {
            pCoeff[i] = @(0);
        }
    }
    
    return pCoeff;
}

/**
 * Find the frequency response of the signal synthesized with above LPC coefficients
 * calculated in 5 Hz intervals, amplitude is [0,1]
 */
- (NSArray *)synthesizedFrequencyResponse
{
    double *pCoeff = (double *)(malloc((ORDER + 1) * sizeof(double)));
    NSArray *lpcCoefficients = [self lpcCoefficients];
    for (int i=0; i<=ORDER; i++) {
        pCoeff[i] = ((NSNumber *)lpcCoefficients[i]).doubleValue;
    }
    
    // Now we find frequency response of the inverse of the predictor filter
    long degIdx, k;
    
    // A few variable used in plotting of H(w).
    double omega, realHw, imagHw;
    
    NSMutableArray *freqResponse = [NSMutableArray arrayWithCapacity:300];
    
    for ( degIdx=0; degIdx < 300; degIdx++) {
        omega = degIdx * M_PI / 330.0;
        realHw = 1.0;
        imagHw = 0.0;
        
        for (k = 1 ; k <= ORDER ; k++) {
            realHw = realHw + pCoeff[k] * cos(k * omega);
            imagHw = imagHw - pCoeff[k] * sin(k * omega);
        }
        
        freqResponse[degIdx] = @(20*log10(1.0 / sqrt(realHw * realHw + imagHw * imagHw)));
    }
    
    return freqResponse;
}


/**
 * Heart of Laguerre algorithm. Solved the polynomial equation of a certain order.
 * This functions is called repeatedly to find all the complex roots one by one.
 */
+ (_Complex double)laguer:(_Complex double *)a currentOrder:(int)m
{
    int iter , j;
    double abx , abp , abm , err;
    _Complex double dx , x , x1 , b , d , f , g , h , sq , gp , gm , g2;
    static float frac[MR+1] = {0.0,0.5,0.25,0.75,0.13,0.38,0.62,0.88,1.0};
    
    x = 0.0 + 0.0 * I;
    
    for (iter = 1 ; iter <= MAXIT ; iter++) {
        b = a[m];
        err = cabs(b);
        d = f = 0.0 + 0.0 * I;
        abx = cabs(x);
        
        for (j = m - 1 ; j >= 0 ; j--) {
            f = x * f + d;
            d = x * d + b;
            b = x * b + a[j];
            err = cabs(b) + abx * err;
        }
        
        err *= EPSS;
        if (cabs(b) <= err) {    // error is small, return x even if iterations are not exhausted
            return x;
        }
        
        g = d / b;
        g2 = g * g;
        h = g2 - f / b;
        sq = csqrt( (m-1) * (m * h - g2) );
        gp = g + sq;
        gm = g - sq;
        abp = cabs(gp);
        abm = cabs(gm);
        if (abp < abm)
        {
            gp=gm;
        }
        
        dx = ((MAX(abp,abm) > 0.0 ? (m + 0.0 * I) / gp
               : (1+abx) * (cos((float)iter) + sin((float)iter) * I)));
        x1 = x - dx;
        if (creal(x) == creal(x1) && cimag(x) == cimag(x1))
        {
            return x;
        }
        
        if (iter % MT) x = x1; else x = x - frac[iter/MT] * dx;
    }
    return x;
}

/**
 * Following function implement Laguerre root finding algorithm. It uses a lot of
 * complex variables and operations of complex variables. It does not implement
 * root polishing so answers are not very accurate.
 * Input: pCoeff
 */
+ (double *)findFormants:(_Complex double*)a
{
    // Allocate space for complex roots
    _Complex double *roots = (_Complex double *)(malloc((ORDER+1) * sizeof(_Complex double)));
    
    int j , jj;
    _Complex double x, b, c;
    
    _Complex double *ad = (_Complex double *)(malloc((ORDER+1) * sizeof(_Complex double)));
    
    for (j = 0 ; j <= ORDER ; j++)
    {
        ad[j] = a[j];
    }
    
    for (j = ORDER ; j >= 1 ; j--) {
        x = [SpeechAnalyzer laguer:ad currentOrder:j];
        
        // If imaginary part is very small, ignore it
        if (fabs(cimag(x)) <= 2.0*EPS*fabs(creal(x)))
        {
            x = creal(x) + 0.0 * I;
        }
        
        roots[j] = x;
        
        // Perform forward deflation. Divide by the factor of the root found above
        b = ad[j];
        for (jj = j-1 ; jj >= 0 ; jj--) {
            c = ad[jj];
            ad[jj] = b;
            b = x * b + c;
        }
    }
    
    // Find real-frequencies corresponding to all roots and fill the array.
    
    // Allocate space for real-world frequencies
    double *formantFrequencies = (double *)(malloc((ORDER+1) * sizeof(double)));
    
    for (int dummo=0; dummo<=ORDER; dummo++) {
        formantFrequencies[dummo] = 0.0;
    }
    
    for (int dummo=0; dummo<=ORDER; dummo++) {
        formantFrequencies[dummo] = 5512.5 * carg(roots[dummo]) / M_PI;
    }
    
    return formantFrequencies;
}

/**
 * Finds the first four formants and cleans out negatives, and other problems
 * Return is array of formants in Hz
 */
- (NSArray *)findCleanFormants
{
    if (!self.cleanFormantsCached) {
        short int *decimatedBuffer = (short int *)self.int16SamplesDecimated.bytes;
        long decimatedBufferSamples = self.int16SamplesDecimated.length / sizeof(short int);
        
        long i, j, dummo;
        
        double *formantFrequencies;
        double dummyFrequency;
        
        // Here we find complex roots from the LP filter to find formant frequencies.
        // First find LPCoefficients by repeating a few blocks given above.
        
        // Find ORDER+1 autocorrelation coefficient
        double *Rxx = (double *)(malloc((ORDER + 1) * sizeof(double)));
        double *pCoeff = (double *)(malloc((ORDER + 1) * sizeof(double)));
        
        for (int delayIdx = 0; delayIdx <= ORDER; delayIdx++) {
            double corrSum = 0;
            for (int dataIdx = 0; dataIdx < (decimatedBufferSamples - delayIdx); dataIdx++) {
                corrSum += (decimatedBuffer[dataIdx] * decimatedBuffer[dataIdx + delayIdx]);
            }
            
            Rxx[delayIdx] = corrSum;
        }
        
        // Now solve for the predictor coefficients
        double pError = Rxx[0];                             // initialise error to total power
        pCoeff[0] = 1.0;                                    // first coefficient must be = 1
        
        // for each coefficient in turn
        for (int k = 1 ; k <= ORDER ; k++) {
            
            // find next reflection coeff from pCoeff[] and Rxx[]
            double rcNum = 0;
            for (int i = 1 ; i <= k ; i++)
            {
                rcNum -= pCoeff[k-i] * Rxx[i];
            }
            
            pCoeff[k] = rcNum/pError;
            
            // perform recursion on pCoeff[]
            for (int i = 1 ; i <= k/2 ; i++) {
                double pci  = pCoeff[i] + pCoeff[k] * pCoeff[k-i];
                double pcki = pCoeff[k-i] + pCoeff[k] * pCoeff[i];
                pCoeff[i] = pci;
                pCoeff[k-i] = pcki;
            }
            
            // calculate residual error
            pError = pError * (1.0 - pCoeff[k]*pCoeff[k]);
            
        }
        
        // Now work with a lot of complex variables to find complex roots of LPC filter.
        // These roots will give us formant frequencies.
        
        _Complex double *compCoeff = (_Complex double *)(malloc((ORDER + 1) * sizeof(_Complex double)));
        
        // Transfer pCoeff (real-valued) to compCoeff (complex-valued).
        for (dummo=0; dummo <= ORDER; dummo++) {
            compCoeff[dummo] = pCoeff[ORDER - dummo] + 0.0 * I;
        }
        
        // Formant frequencies are computed in a separate function.
        formantFrequencies = [SpeechAnalyzer findFormants:compCoeff];
        
        //Now clean formant frequencies. Remove all that are negative, < 50 Hz, or > (Fs/2 - 50).
        for (dummo = 1; dummo <= ORDER; dummo++) {
            if (formantFrequencies[dummo] > (5512.5 - 50.0))  formantFrequencies[dummo] = 5512.5;
            if (formantFrequencies[dummo] < 50.0)  formantFrequencies[dummo] = 5512.5;
        }
        
        // Now sort formant frequencies. Simple in-place bubble sort.
        for (i = 1 ; i <= ORDER ; i++) {
            for (j = i ; j <= ORDER ; j++) {
                if (formantFrequencies[i] > formantFrequencies[j]) {
                    dummyFrequency = formantFrequencies[i];
                    formantFrequencies[i] = formantFrequencies[j];
                    formantFrequencies[j] = dummyFrequency;
                }
            }
        }
        
        // Now list first 8 sorted frequencies.
        for (dummo = 1; dummo <= 8; dummo++) {
            NSLog(@"Format frequency for index %ld is %5.0f",dummo, formantFrequencies[dummo]);
        }
        
        // Print a blank line
        NSLog(@" ");
        
        self.cleanFormantsCached = @[@(formantFrequencies[1]),
                                     @(formantFrequencies[2]),
                                     @(formantFrequencies[3]),
                                     @(formantFrequencies[4])];
        
        // Free memory allocated by our routine
        free(Rxx);
        free(pCoeff);
        free(compCoeff);
    }
    return self.cleanFormantsCached;
}





@end

File: Formant Analyzer/Formant Analyzer/SpeechAnalyzer.swift
//
//  SpeechAnalyzer.swift
//  FormantPlotter
//
//  Created by William Entriken on 1/22/16.
//  Copyright © 2016 William Entriken. All rights reserved.
//

import Foundation

/// Analyzes speech signals to extract formants and related properties.
class SpeechAnalyzer {
    /// Individual audio samples.
    let samples: [Int16]
    
    /// The rate in Hz.
    let sampleRate: Int
    
    /// Human formants are < 5 kHz, so we do not need signal information above 10 kHz.
    lazy var decimationFactor: Int = {
        max(1, sampleRate / 10000)
    }()
    
    /// The part of `samples` which has a strong signal.
    lazy var strongPart: CountableRange<Int> = {
        SpeechAnalyzer.findStrongPartOfSignal(samples, withChunks: 300, sensitivity: 0.1)
    }()
    
    /// The part of `samples` which is a vowel utterance.
    lazy var vowelPart: CountableRange<Int> = {
        strongPart.truncatedTails(byPortion: 0.15)
    }()
    
    /// The vowel part of `samples` decimated by `decimationFactor`.
    private lazy var vowelSamplesDecimated: [Int16] = {
        let range = vowelPart
        return samples[range].decimated(by: decimationFactor)
    }()
    
    /// Linear prediction coefficients of the vowel signal.
    lazy var estimatedLpcCoefficients: [Double] = {
        SpeechAnalyzer.estimateLpcCoefficients(samples: vowelSamplesDecimated, sampleRate: sampleRate / decimationFactor, modelLength: 10)
    }()
    
    /// Synthesized frequency response for the estimated LPC coefficients.
    /// - Returns: The response at frequencies 0, 15, ... Hz, with the first index (identity) as 1.0.
    lazy var synthesizedFrequencyResponse: [Double] = {
        let frequencies = Array(stride(from: 0, to: sampleRate / decimationFactor / 2, by: 15))
        return SpeechAnalyzer.synthesizeResponseForLPC(estimatedLpcCoefficients, withRate: sampleRate / decimationFactor, atFrequencies: frequencies)
    }()
    
    /// Finds at least the first four formants in the range for estimating human vowel pronunciation.
    /// - Returns: Formants in Hz.
    lazy var formants: [Double] = {
        let complexPolynomial = estimatedLpcCoefficients.map { 0.0.i + $0 }
        let formants = SpeechAnalyzer.findFormants(complexPolynomial, sampleRate: sampleRate / decimationFactor)
        return SpeechAnalyzer.filterSpeechFormants(formants)
    }()
    
    /// Reduces horizontal resolution of `strongPart` for plotting.
    /// - Parameter newSampleCount: The desired number of samples.
    /// - Returns: Downsampled signal values.
    func downsampleStrongPartToSamples(_ newSampleCount: Int) -> [Int16] {
        guard newSampleCount > 0, !strongPart.isEmpty else { return [] }
        let chunkSize = strongPart.count / newSampleCount
        guard chunkSize > 0 else { return Array(samples[strongPart].prefix(newSampleCount)) }
        
        var chunkMaxElements: [Int16] = []
        for chunkStart in stride(from: strongPart.lowerBound, through: strongPart.upperBound - chunkSize, by: chunkSize) {
            let range = chunkStart..<min(chunkStart + chunkSize, strongPart.upperBound)
            chunkMaxElements.append(samples[range].max() ?? 0)
        }
        return chunkMaxElements
    }
    
    /// Reduces horizontal resolution of `samples` for plotting.
    /// - Parameter newSampleCount: The desired number of samples.
    /// - Returns: Downsampled signal values.
    func downsampleToSamples(_ newSampleCount: Int) -> [Int16] {
        guard newSampleCount > 0 else { return [] }
        guard newSampleCount < samples.count else { return samples }
        let chunkSize = samples.count / newSampleCount
        guard chunkSize > 0 else { return Array(samples.prefix(newSampleCount)) }
        
        var chunkMaxElements: [Int16] = []
        for chunkStart in stride(from: samples.startIndex, through: samples.endIndex - chunkSize, by: chunkSize) {
            let range = chunkStart..<min(chunkStart + chunkSize, samples.endIndex)
            chunkMaxElements.append(samples[range].max() ?? 0)
        }
        return chunkMaxElements
    }
    
    /// Creates an analyzer with given 16-bit PCM samples.
    /// - Parameters:
    ///   - data: The raw PCM data.
    ///   - rate: The sample rate in Hz.
    init(int16Samples data: Data, withFrequency rate: Int) {
        samples = data.withUnsafeBytes { buffer in
            Array(buffer.bindMemory(to: Int16.self))
        }
        sampleRate = rate
    }
    
    /// Analyzes a signal to find the significant part.
    /// - Parameters:
    ///   - signal: The input signal.
    ///   - numChunks: Number of chunks to divide the signal into.
    ///   - factor: Sensitivity factor for energy threshold.
    /// - Returns: The range of the strong signal.
    class func findStrongPartOfSignal(_ signal: [Int16], withChunks numChunks: Int, sensitivity factor: Double) -> CountableRange<Int> {
        guard !signal.isEmpty, numChunks > 0 else { return 0..<0 }
        let chunkSize = max(1, signal.count / numChunks)
        var chunkEnergies: [Double] = []
        var maxChunkEnergy: Double = 0
        
        for chunkStart in stride(from: signal.startIndex, through: signal.endIndex - chunkSize, by: chunkSize) {
            let range = chunkStart..<min(chunkStart + chunkSize, signal.endIndex)
            let chunkEnergy = signal[range].reduce(0.0) { $0 + Double($1) * Double($1) }
            maxChunkEnergy = max(maxChunkEnergy, chunkEnergy)
            chunkEnergies.append(chunkEnergy)
        }
        
        let firstSelectedChunk = chunkEnergies.firstIndex { $0 > maxChunkEnergy * factor } ?? 0
        let lastSelectedChunk = chunkEnergies.reversed().firstIndex { $0 > maxChunkEnergy * factor }.map { chunkEnergies.count - 1 - $0 } ?? chunkEnergies.count - 1
        return firstSelectedChunk * chunkSize..<min((lastSelectedChunk + 1) * chunkSize, signal.count)
    }
    
    /// Estimates LPC polynomial coefficients from the signal using Levinson-Durbin recursion.
    /// - Parameters:
    ///   - samples: The input samples.
    ///   - rate: The sample rate in Hz.
    ///   - modelLength: The number of coefficients.
    /// - Returns: Autocorrelation coefficients for an all-pole model.
    class func estimateLpcCoefficients(samples: [Int16], sampleRate rate: Int, modelLength: Int) -> [Double] {
        guard samples.count > modelLength else {
            return [Double](repeating: 1, count: modelLength + 1)
        }
        
        var correlations: [Double] = []
        var coefficients: [Double] = []
        var modelError: Double
        
        for delay in 0...modelLength {
            var correlationSum = 0.0
            for sampleIndex in 0..<(samples.count - delay) {
                correlationSum += Double(samples[sampleIndex]) * Double(samples[sampleIndex + delay])
            }
            correlations.append(correlationSum)
        }
        
        modelError = correlations[0]
        guard modelError != 0 else { return [Double](repeating: 1, count: modelLength + 1) }
        coefficients.append(1.0)
        
        for delay in 1...modelLength {
            var rcNum = 0.0
            for i in 1...delay {
                rcNum -= coefficients[delay - i] * correlations[i]
            }
            coefficients.append(rcNum / modelError)
            
            for i in stride(from: 1, through: delay / 2, by: 1) {
                let pci = coefficients[i] + coefficients[delay] * coefficients[delay - i]
                let pcki = coefficients[delay - i] + coefficients[delay] * coefficients[i]
                coefficients[i] = pci
                coefficients[delay - i] = pcki
            }
            
            modelError *= 1.0 - coefficients[delay] * coefficients[delay]
        }
        return coefficients
    }
    
    /// Synthesizes the frequency response for the estimated LPC coefficients.
    /// - Parameters:
    ///   - coefficients: The LPC model coefficients.
    ///   - samplingRate: The sampling frequency in Hz.
    ///   - frequencies: The frequencies to evaluate.
    /// - Returns: The response from 0 to 1 for each frequency.
    class func synthesizeResponseForLPC(_ coefficients: [Double], withRate samplingRate: Int, atFrequencies frequencies: [Int]) -> [Double] {
        var response: [Double] = []
        for frequency in frequencies {
            let radians = Double(frequency) / Double(samplingRate) * Double.pi * 2
            var sum: Complex<Double> = 0.0 + 0.0.i
            for (index, coefficient) in coefficients.enumerated() {
                sum += Complex<Double>(abs: coefficient, arg: Double(index) * radians)
            }
            response.append(20 * log10(1.0 / sum.abs))
        }
        return response
    }
    
    /// Finds one root of a complex polynomial using Laguerre's method.
    /// - Parameters:
    ///   - polynomial: The polynomial coefficients.
    ///   - guess: The initial guess for the root.
    /// - Returns: The computed root.
    class func laguerreRoot(_ polynomial: [Complex<Double>], initialGuess guess: Complex<Double> = 0.0 + 0.0.i) -> Complex<Double> {
        let m = polynomial.count - 1
        let MR = 8
        let MT = 10
        let maximumIterations = MR * MT
        let EPSS = 1.0e-7
        
        var x = guess
        let frac = [0.0, 0.5, 0.25, 0.75, 0.125, 0.375, 0.625, 0.875, 1.0]
        
        for iteration in 1...maximumIterations {
            var b = polynomial[m]
            var err = b.abs
            var d: Complex<Double> = 0.0 + 0.0.i
            var f: Complex<Double> = 0.0 + 0.0.i
            let abx = x.abs
            
            for j in stride(from: m - 1, through: 0, by: -1) {
                f = x * f + d
                d = x * d + b
                b = x * b + polynomial[j]
                err = b.abs + abx * err
            }
            err *= EPSS
            
            if b.abs < err {
                return x
            }
            
            let g = d / b
            let g2 = g * g
            let h = g2 - 2.0 * f / b
            let sq = sqrt((Double(m) - 1) * (Double(m) * h - g2))
            var gp = g + sq
            let gm = g - sq
            let abp = gp.abs
            let abm = gm.abs
            if abp < abm {
                gp = gm
            }
            let dx = max(abp, abm) > 0.0 ? Double(m) / gp : (1 + abx) * (cos(Double(iteration)) + sin(Double(iteration)).i)
            let x1 = x - dx
            
            if x == x1 {
                return x
            }
            
            x = iteration % MT > 0 ? x1 : x - frac[iteration / MT] * dx
        }
        NSLog("Too many iterations in Laguerre, returning zero")
        return 0 + 0.i
    }
    
    /// Finds the roots of a complex polynomial using Laguerre's method.
    /// - Parameters:
    ///   - polynomial: The polynomial coefficients.
    ///   - rate: The sample rate in Hz.
    /// - Returns: Formant frequencies in Hz.
    class func findFormants(_ polynomial: [Complex<Double>], sampleRate rate: Int) -> [Double] {
        var roots: [Complex<Double>] = []
        var deflatedPolynomial = polynomial
        let modelOrder = polynomial.count - 1
        
        for j in stride(from: modelOrder, through: 1, by: -1) {
            var root = laguerreRoot(deflatedPolynomial)
            if abs(root.imag) < 2.0e-6 * abs(root.real) {
                root.imag = 0.0
            }
            roots.append(root)
            
            var b = deflatedPolynomial[j]
            for jj in stride(from: j - 1, through: 0, by: -1) {
                let c = deflatedPolynomial[jj]
                deflatedPolynomial[jj] = b
                b = root * b + c
            }
        }
        
        let polishedRoots = roots.map { laguerreRoot(polynomial, initialGuess: $0) }
        let formantFrequencies = polishedRoots.map { $0.arg * Double(rate) / Double.pi / 2 }
        return formantFrequencies.sorted()
    }
    
    /// Filters formants to ensure they are valid for human speech.
    /// - Parameter formants: The input formants in Hz.
    /// - Returns: At least four filtered formants in Hz.
    class func filterSpeechFormants(_ formants: [Double]) -> [Double] {
        let MIN_FORMANT = 50.0
        let MAX_FORMANT = 5000.0
        let MIN_DISTANCE = 10.0
        
        var editedFormants = formants.sorted().filter { $0 >= MIN_FORMANT && $0 <= MAX_FORMANT }
        var done = false
        while !done {
            done = true
            for (index, formantA) in editedFormants.enumerated() {
                guard index < editedFormants.count - 1 else { continue }
                let formantB = editedFormants[index + 1]
                if abs(formantA - formantB) < MIN_DISTANCE {
                    let newFormant = (formantA + formantB) / 2
                    editedFormants.remove(at: index + 1)
                    editedFormants[index] = newFormant
                    editedFormants.sort()
                    done = false
                    break
                }
            }
        }
        
        while editedFormants.count < 4 {
            editedFormants.append(9999.0)
        }
        
        return editedFormants
    }
}

/// Truncates the tails of a range by a portion of its length.
extension CountableRange where Bound.Stride == Int {
    /// Shrinks range by `portion` of its length from each side.
    /// - Parameter portion: A fraction in the range 0...0.5.
    func truncatedTails(byPortion portion: Double) -> CountableRange<Bound> {
        let start = lowerBound.advanced(by: Int((portion * Double(count)).rounded()))
        let end = lowerBound.advanced(by: Int(((1 - portion) * Double(count)).rounded()))
        return start..<end
    }
}

extension ArraySlice {
    /// Selects the first of every `step` items.
    /// - parameter step: the decimation factor.
    /// - returns: the decimated array.
    func decimated(by step: Int) -> [Element] {
        guard step > 0 else { return [] }
        return Swift.stride(from: startIndex, to: endIndex, by: step).map { self[$0] }
    }
}
